### ISSUE TITLE
Node not switching

### ISSUE THREAD

**Issue by selrond** [Schniz/fnm#583]
I've installed `fnm` on mac using homebrew, and put `eval "$(fnm env)"` inside my `.zshrc`.
It appears to work, but when I run `fnm use <version>`, and test that version using `node -v`, output versions are different, as if `fnm` didn't switch it at all. 

It only works after I re-source `.zshrc`. I guess it's not supposed to work that way...

Is there anything I have to do to make it work that I overlooked?

**Comment by Schniz** [comment]
hmmm

can you please share your `$PATH` env var? :)

**Comment by selrond** [comment]
`/usr/local/sbin:/var/folders/_r/slz2zc7s2hs59_8rtc3lww900000gn/T/fnm_multishells/13700_1637656413932/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/MacGPG2/bin:/Library/Apple/usr/bin
`

**Comment by selrond** [comment]
It's weird, because when I launch VSCode from the terminal, (so it uses the proper node version) and check the node version from within it (using its integrated terminal), it outputs proper version.

Only the current terminal session seems to not be updating

**Comment by selrond** [comment]
<img width="764" alt="image" src="https://user-images.githubusercontent.com/6603389/143060550-0230b8d5-a866-4f4e-9455-6563687b3bb0.png">

**Comment by Schniz** [comment]
try running `rehash` after `fnm use 14` please :) just to make sure that it works

**Comment by selrond** [comment]
yep, that seems to fix it - do I have to run it every time after changing the version though?

<img width="181" alt="image" src="https://user-images.githubusercontent.com/6603389/143072092-b7fb4bb0-2ceb-411f-97dc-e6e293dadd1a.png">

**Comment by Schniz** [comment]
I need to check why this is happening. It did not happen to me and I'm using Zsh, maybe that's a newer version and we need to fix fnm somehow.

**Comment by selrond** [comment]
sure, if I can be of any help, let me know

**Comment by selrond** [comment]
@Schniz it appears to be fixed since https://github.com/Schniz/fnm/releases/tag/v1.28.1

I've taken a look at the PR, and although I don't know a thing when it comes to Rust, maybe [this line](https://github.com/Schniz/fnm/pull/576/files#diff-bd6bf7f14d05190b3bed384ceb3a6236f1c67a637890327d6746dda827352a0cL18) is related? (based on `hashmap` keyword)

Anyway, let's close this and I'll report back if I'll encounter it again

**Comment by Schniz** [comment]
weird, it should not affect 🤔 for some reason, rehashing is required _sometimes_. I don't know why, but I have an idea how to work it out and fix other issues with the same fix :)

closing for now :)

**Comment by selrond** [comment]
@Schniz I’ve started to experience this behavior again... 😕

**Comment by alexeyten** [comment]
Could you show `echo $PATH` and `hash -t node`?

**Comment by selrond** [comment]
`PATH`:

```sh
/usr/local/sbin:/Users/selrond/Library/Caches/fnm_multishells/85078_1644225259663/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Apple/usr/bin
```

`hash -t node` throws `hash: bad option: -t`

**Comment by alexeyten** [comment]
Oh, you use `zsh`. Then it's  `hash -L | grep node=` command.

**Comment by selrond** [comment]
thx, here it is:

`hash node=/Users/selrond/Library/Caches/fnm_multishells/4390_1644223494403/bin/node`

**Comment by selrond** [comment]
@alexeyten do you know of a solution to this problem?

**Comment by alexeyten** [comment]
Nope. I don't use zsh and macos.

**Comment by selrond** [comment]
> I need to check why this is happening. It did not happen to me and I'm using Zsh, maybe that's a newer version and we need to fix fnm somehow.

@Schniz Any chance I could help to pin this issue down?

**Comment by Schniz** [comment]
There’s a hack I researched. Rehashing on zsh automatically. I’ll open a PR tomorrow and then you can test it before merging. Wdyt?

**Comment by selrond** [comment]
I’m no expert in zsh, so whatever works 😄 

I’ve noticed a strange(?) behavior - once I run `rehash` once in a terminal session, it seems to be working (tried switching versions back and forth and running `node -v`)

**Comment by selrond** [comment]
Also, [this](https://github.com/Schniz/fnm/issues/583#issuecomment-976457223) is no longer a case:

> when I launch VSCode from the terminal, (so it uses the proper node version) and check the node version from within it (using its integrated terminal), it outputs proper version

`where node` inside its terminal returns this:

```sh
/opt/homebrew/bin/node
/Users/selrond/Library/Caches/fnm_multishells/3981_1644995919375/bin/node
/opt/homebrew/bin/node
```

**Comment by Schniz** [comment]
I do not use VSCode so it'll be hard for me to debug & verify. I don't know what it's running in the integrated shell, etc.

`rehash` makes sense, this is why we're doing it 🤔 
https://github.com/Schniz/fnm/blob/1acb2892c502c73e4acd0469410f1b8ffa687678/src/shell/zsh.rs#L26-L28

**Comment by Schniz** [comment]
my idea was to call `rehash` before calls to `node` or before any call (yet this feels... not good 😁)

**Comment by selrond** [comment]
I’ve solved the VSCode problem by not running its terminal as a login shell, so that’s out of question.

Only the rehash-needed problem remains

**Comment by selrond** [comment]
I’ve "solved" it for now by uninstalling system nodejs (installed via homebrew).
Not exactly happy with the solution, but it seems to work.
Really weird, I’ve never came across `rehash` previously, and I’ve been tinkering with terminal setup for some years now...

**Comment by peeeyow** [comment]
I have same issue here. I'm using `zsh 5.8.1` and `fnm 1.29.2`. The problem only appears when the system nodejs is involved. When I currently use the system node it works perfectly fine. But when I change the node version (either manually `fnm use` or when I cd to a directory with `.node-version`) it node version does not work. `which node` shows proper path (like `/tmp/fnm_multishells/7982_1649893314783/bin/node`)  but node still uses the system node (`/usr/bin/node`).

**Comment by boomikuma** [comment]
I also have the same problem. When switching node version from system to v14, the version doesn't change unless `rehash` is called. However, after calling rehash, switching back to system then v14 will switch the version.


<img width="680" alt="177051111-700919db-2a2a-4a1f-922b-4546b2f958b6" src="https://user-images.githubusercontent.com/37561295/177051182-e45df4f4-bd6d-441a-a81f-7b1f48b0f486.png">

**Comment by ouchxp** [comment]
I have tried the solution [here](https://stackoverflow.com/questions/36543707/why-does-zsh-ignore-the-path-entry-order), simply adding `setopt nohashcmds` to `.zshrc`  and it worked for me.

**Comment by Crazy-Hopper** [comment]
> I have tried the solution [here](https://stackoverflow.com/questions/36543707/why-does-zsh-ignore-the-path-entry-order), simply adding `setopt nohashcmds` to `.zshrc` and it worked for me.

For the sake of completeness, here is the bash analogue: `set +o hashall` or shorter `set +h`

**Comment by braineo** [comment]
hi @Schniz, will calling `rehash` after `fnm use` automatically will fix this problem?

**Comment by hyrious** [comment]
I end up rewriting my .zshrc like this:

```zsh
# fnm
eval "$(fnm env --shell zsh)"
# --use-on-cd + rehash, https://github.com/Schniz/fnm/issues/583
autoload -U add-zsh-hook
_fnm_autoload_hook () {
  if [[ -f .node-version || -f .nvmrc || -f package.json ]]; then
    fnm use --silent-if-unchanged
    rehash  # <--
  fi
}
add-zsh-hook chpwd _fnm_autoload_hook && _fnm_autoload_hook
```
